src/App/Fossa/LicenseScanner.hs

```diff
@@ -0,0 +1,172 @@
+{-# LANGUAGE RecordWildCards #-}
+
+module App.Fossa.LicenseScanner (
+  licenseScanSourceUnit,
+  licenseNoScanSourceUnit,
+) where
+
+import App.Fossa.ArchiveUploader (
+  VendoredDependency (..),
+  arcToLocator,
+  duplicateFailureBundle,
+  duplicateNames,
+  forceVendoredToArchive,
+ )
+import App.Fossa.EmbeddedBinary (ThemisBins, withThemisAndIndex)
+import App.Fossa.FossaAPIV1 qualified as Fossa
+import App.Fossa.RunThemis (
+  execThemis,
+ )
+import Control.Effect.Diagnostics (Diagnostics, ToDiagnostic (renderDiagnostic), context, fatalText, fromMaybe)
+import Control.Effect.Lift (Has, Lift)
+import Control.Effect.StickyLogger (StickyLogger, logSticky)
+import Control.Monad (unless, void)
+import Crypto.Hash (Digest, MD5, hash)
+import Data.ByteString qualified as B
+import Data.List.NonEmpty (NonEmpty)
+import Data.List.NonEmpty qualified as NE
+import Data.Maybe (catMaybes)
+import Data.String.Conversion (encodeUtf8, toString)
+import Data.Text.Extra (showT)
+import Effect.Exec (Exec)
+import Effect.Logger (Logger, Pretty (pretty), logError)
+import Fossa.API.Types (
+  ApiOpts,
+  Archive (Archive, archiveName),
+  ArchiveComponents (ArchiveComponents),
+  Organization (organizationId),
+ )
+import Path (Abs, Dir, Path, parseRelDir, (</>))
+import Srclib.Types (
+  LicenseScanType (CliLicenseScanned),
+  LicenseSourceUnit (..),
+  LicenseUnit (..),
+  Locator (..),
+ )
+
+data NoSuccessfulScans = NoSuccessfulScans
+
+instance ToDiagnostic NoSuccessfulScans where
+  renderDiagnostic _ = "No native license scans were successful"
+
+runLicenseScanOnDir ::
+  ( Has Diagnostics sig m
+  , Has (Lift IO) sig m
+  , Has Exec sig m
+  ) =>
+  Path Abs Dir ->
+  m [LicenseUnit]
+runLicenseScanOnDir scanDir = withThemisAndIndex $ themisRunner scanDir
+
+themisRunner ::
+  ( Has (Lift IO) sig m
+  , Has Diagnostics sig m
+  , Has Exec sig m
+  ) =>
+  Path Abs Dir ->
+  ThemisBins ->
+  m [LicenseUnit]
+themisRunner scanDir themisBins = runThemis themisBins scanDir
+
+runThemis :: (Has Exec sig m, Has Diagnostics sig m) => ThemisBins -> Path Abs Dir -> m [LicenseUnit]
+runThemis themisBins scanDir = do
+  context "Running license scan binary" $ execThemis themisBins scanDir
+
+md5 :: B.ByteString -> Digest MD5
+md5 = hash
+
+scanAndUploadVendoredDep ::
+  ( Has Diagnostics sig m
+  , Has (Lift IO) sig m
+  , Has Logger sig m
+  , Has StickyLogger sig m
+  , Has Exec sig m
+  ) =>
+  ApiOpts ->
+  Path Abs Dir ->
+  VendoredDependency ->
+  m (Maybe Archive)
+scanAndUploadVendoredDep apiOpts baseDir vendoredDeps@VendoredDependency{..} = context "compressing and uploading vendored deps" $ do
```

Comment:
[Optional] prefer if we break this function into `scanVendorDeps` + `scanAndUploadVendorDeps`. This was  repl'ing is easier.